### PR TITLE
3 jupyter widget ipywidgets

### LINKED ISSUE #7: Widget: initial implementation
Initial implementation of a view of a widget. 

Epic https://github.com/IceKhan13/purplecaffeine/issues/3

### PR DESCRIPTION
**Pull request by mentesniker**
### Summary

Widget: ipywidget draft implementation

### Details and comments

Closes https://github.com/IceKhan13/purplecaffeine/issues/7

### CODE REVIEW

File: docs/guides/02_widget.ipynb
Comment (mickahell) [inline docs/guides/02_widget.ipynb:1]: Can you rename the API guide to be the 4th :)
[guide : docs/guides/02_setup_api.md](https://github.com/IceKhan13/purplecaffeine/blob/main/docs/guides/02_setup_api.md)

File: docs/guides/trials/007d7c6b-3fcf-44a4-8d5d-2105dd43a21c.json
Comment (mickahell) [inline docs/guides/trials/007d7c6b-3fcf-44a4-8d5d-2105dd43a21c.json:1]: Why do you keep this file ?

File: requirements-dev.txt [hunk 1]
```diff
@@ -9,3 +9,5 @@ nbconvert==7.4.0
 nbformat==5.8.0
 testcontainers==3.7.1
 qiskit_optimization==0.5.0
+pandas
+matplotlib
```
Comment (mickahell) [inline requirements-dev.txt]: both should be in the classical requirements.txt not here :
https://github.com/IceKhan13/purplecaffeine/blob/3-jupyter-widget-ipywidgets/client/requirements.txt

Comment (mickahell) [reply requirements-dev.txt]: Also miss `ipywidgets` in the requirements.txt as well

Comment (mickahell) [reply requirements-dev.txt]: Also can you put a version (latest if works) :)

File: docs/requirements-doc.txt [hunk 2]
```diff
@@ -18,3 +18,5 @@ qiskit-terra>=0.21.1
 sphinx-copybutton>=0.5.0
 qiskit-ibm-runtime==0.9.3
 boto3==1.26.133
+pandas
+matplotlib
```
Comment (mickahell) [inline docs/requirements-doc.txt]: Can you put a version (same as the one for requirements.txt) :)

File: docs/guides/02_widget.ipynb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "1dc0afee-000c-4ce2-8454-729c1085e74a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "local_storage = LocalStorage(\"./\")\n",
```
Comment (mickahell) [inline docs/guides/02_widget.ipynb]: Put this after the import for better reading ;)

**Review by mickahell — CHANGES_REQUESTED** [review]
Hi Luis :) I just put some no related feature review. So you can have a first review for basic stuffs and no need to worry then 🚀

**Comment by mickahell** [comment]
Also, quick question. What about these branchs `3-jupyter-widget` and `chore/widget` ? Do you still need them for copy and paste stuffs ?

**Comment by mentesniker** [comment]
> Also, quick question. What about these branchs `3-jupyter-widget` and `chore/widget` ? Do you still need them for copy and paste stuffs ?

Yes we can delete it

File: docs/guides/02_widget.ipynb
Comment (mentesniker) [reply docs/guides/02_widget.ipynb:1]: done

File: client/purplecaffeine/widget.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+    def render_trails_list(self):
+        """
+        Render the list of trail buttons.
+        We render one button per each trial
+        """
+        buttons = []
+        for trial in self.current_page_trials:
+            button = widgets.Button(
+                description=f"{trial.name} | {trial.uuid}"[:30],
```
Comment (mickahell) [inline client/purplecaffeine/widget.py:115]: Maybe do not show the uuid in the title, that's not very user friendly

File: client/purplecaffeine/widget.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        metrics.layout = Layout(overflow="scroll", max_height="300px")
+        circuits = widgets.Output()
+        with circuits:
+            for name, circuit in self.selected_trial.circuits:
+                print(name)
+                print(circuit)
+        circuits.layout = Layout(overflow="scroll", max_height="300px")
+
+        tab = widgets.Tab(children=[info, metrics, circuits])
+        tab.titles = ["Info", "Metrics", "Circuits"]
```
Comment (mickahell) [inline client/purplecaffeine/widget.py]: Why do you only have 3 tabs ?
Trial has :
- metrics
- parameters
- circuits
- operators
- artifacts
- arrays

File: client/tests/test_widget.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
+"""Tests for Trial."""
```
Comment (mickahell) [inline client/tests/test_widget.py]:
```suggestion
"""Tests for Widget."""
```

File: client/tests/test_widget.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        widget = Widget(self.local_storage)
+        self.assertEqual(
+            widget.render_trails_list().children[0].tooltip,
+            self.local_storage.list()[0].uuid,
+        )
+        self.assertEqual(
+            widget.render_trails_list().children[0].tooltip, widget.selected_trial.uuid
+        )
+
+    def add_n_trials(self, number_of_trials):
```
Comment (mickahell) [inline client/tests/test_widget.py:87]: Move the function on top, for better reading :)

**Review by mickahell — CHANGES_REQUESTED** [review]
Couple of little things to adapt and then for me gonna be ok for launch 🚀

File: client/purplecaffeine/widget.py (hunk 5 above)
Comment (IceKhan13) [reply client/purplecaffeine/widget.py]: this is MVP info will have parameters, title, description; metrics -> metrics, circuit -> circuits. Operators, artifacts, texts and arrays will be added later

File: client/purplecaffeine/widget.py (hunk 4 above)
Comment (IceKhan13) [reply client/purplecaffeine/widget.py:115]: no, it's ok. Sometimes this information is needed on high level

**Comment by IceKhan13** [comment]
Ok, I'm checking PR and have following notes:
- plots for metrics are no longer supported, we should bring it back
- circuits are just printed, while it should be plotted using MPL
- pagination is not working properly and stops after 1 page
- there is a strange mix of bootstrap and original jupyter widget styles, we should pick one and stick to it. Probably bootstrap will go
- widget is not responsive on wide and narrow screens it is not showed properly
- texts, arrays and operators are not in a widget. We should add them

**Comment by IceKhan13** [comment]
I will take it from here! Thank you @mentesniker !

**Comment by IceKhan13** [comment]
Widget updates

Screenshot 2023-06-19 at 3 19 26 PM
Screenshot 2023-06-19 at 3 19 30 PM
Screenshot 2023-06-19 at 3 19 33 PM
Screenshot 2023-06-19 at 3 19 36 PM
Screenshot 2023-06-19 at 3 19 40 PM

File: client/tests/test_widget.py (hunk 7 above)
Comment (IceKhan13) [reply client/tests/test_widget.py:87]: I'm too lazy :) next time :)

**Review by IceKhan13 — APPROVED** [review]
🕺

**Review by mickahell — COMMENTED** [review]
Yeaaaah looks good to me 🚀

**Review by mickahell — APPROVED** [review]

**Review by IceKhan13 — APPROVED** [review]

**Pull request merged**